### PR TITLE
Add support for nsjail compiler executions

### PR DESCRIPTION
**Pull request by mattgodbolt**
Work in progress, but thought I'd make this PR up early for comments and "zomg no don't do that" etc

### CODE REVIEW

File: lib/exec.js [hunk 1]
```diff
@@ -440,6 +467,9 @@ function executeNone(command, args, options) {
 const executeDispatchTable = {
     none: executeNone,
     firejail: executeFirejail,
+    nsjail: (command, args, options) => needsWine(command)
```
Comment (apmorton) [inline lib/exec.js:470]: personal preference, but I would rather just have:
```javascript
if (needsWine(command))
    return executeFirejail(command, args, options);
```
at the top of `executeNsjail`.

File: etc/nsjail/execute.cfg [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+time_limit: 0
+max_cpus: 1
+
+log_level: INFO
+
+rlimit_as_type: INF
+rlimit_cpu_type: SOFT
+rlimit_fsize: 1024 # 1GB
+rlimit_nofile: 20
```
Comment (apmorton) [inline etc/nsjail/execute.cfg:14]: our firejail config does not limit file handles for execution, and 20 file handles seems... optimistic.
I am willing to bet many compilers open significantly more files at once.

File: etc/nsjail/execute.cfg [hunk 3]
```diff
@@ -0,0 +1,137 @@
+name: "compiler explorer compilation/tool sandbox"
+
+mode: ONCE
+hostname: "ce"
+
+time_limit: 0
+max_cpus: 1
```
Comment (apmorton) [inline etc/nsjail/execute.cfg:7]: we probably are better off removing this, see comment below on `cgroup_cpu_ms_per_sec`

File: etc/nsjail/execute.cfg [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+# must run following as root during system startup
+# cgcreate -a ubuntu:ubuntu -g memory,pids,cpu,net_cls:ce-compile
+cgroup_mem_parent: "ce-compile"
+cgroup_pids_parent: "ce-compile"
+cgroup_net_cls_parent: "ce-compile"
+cgroup_cpu_parent: "ce-compile"
+
+cgroup_mem_max: 1342177280 # 1.25 GiB
+cgroup_pids_max: 4
+cgroup_cpu_ms_per_sec: 800
```
Comment (apmorton) [inline etc/nsjail/execute.cfg:33]: this should be set to `1000`

This will let the process group use a single core at 100%, or two cores at 50%

contrasting with max_cpus above, which uses `sched_setaffinity` to pick a random cpu to bind the group to, which can be sub-optimal if multiple compilations or sandbox executions end up picking the same core at once.

Upon further review, for this reason, we probably want to remove the max_cpus from our sandbox config too and rely purely on `cgroup_cpu_ms_per_sec`.

File: etc/nsjail/execute.cfg [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+mount {
```
Comment (apmorton) [inline etc/nsjail/execute.cfg:35]: we are probably going to need to be more liberal with mounts here.

`/usr` <- can't think of a reason to not just give compilers all of this
`/bin` <- wouldn't be surprised if some compilers depend on stuff here
`/etc` <- probably some stuff in here that some compilers may want

Comment (partouf) [reply etc/nsjail/execute.cfg:35]: I agree, this was one of the reasons I was having issues getting this to work with cmake and make.

**Comment by partouf** [comment]
https://godbolt.org/beta/z/5bT8Tc1fq -> pascal broken, it's looking for `as`, so system's binutils need to be accessible

**Comment by partouf** [comment]
```
CMake returns the following error with a basic test:
{
    "buildsteps": [
        {
            "code": 1,
            "okToCache": true,
            "stdout": [
                {
                    "text": "-- Configuring incomplete, errors occurred!"
                },
                {
                    "text": "See also \"/app/CMakeFiles/CMakeOutput.log\"."
                }
            ],
            "stderr": [
                {
                    "text": "[I][2021-04-02T19:04:29+0000] Mode: STANDALONE_ONCE"
                },
                {
                    "text": "[I][2021-04-02T19:04:29+0000] Jail parameters: hostname:'ce', chroot:'', process:'/opt/compiler-explorer/cmake/bin/cmake', bind:[::]:0, max_conns_per_ip:0, time_limit:0, personality:0, daemonize:false, clone_newnet:true, clone_newuser:true, clone_newns:true, clone_newpid:true, clone_newipc:true, clone_newuts:true, clone_newcgroup:true, keep_caps:false, disable_no_new_privs:false, max_cpus:1"
                },
                {
                    "text": "[I][2021-04-02T19:04:29+0000] Mount: '/' flags:MS_RDONLY type:'tmpfs' options:'' dir:true"
                },
                {
                    "text": "[I][2021-04-02T19:04:29+0000] Mount: '/lib' -> '/lib' flags:MS_RDONLY|MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:true"
                },
                {
                    "text": "[I][2021-04-02T19:04:29+0000] Mount: '/usr/lib' -> '/usr/lib' flags:MS_RDONLY|MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:true"
                },
                {
                    "text": "[I][2021-04-02T19:04:29+0000] Mount: '/usr/share/zoneinfo' -> '/usr/share/zoneinfo' flags:MS_RDONLY|MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:true"
                },
                {
                    "text": "[I][2021-04-02T19:04:29+0000] Mount: '/etc/localtime' -> '/etc/localtime' flags:MS_RDONLY|MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:false"
                },
                {
                    "text": "[I][2021-04-02T19:04:29+0000] Mount: '/lib64' -> '/lib64' flags:MS_RDONLY|MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:true mandatory:false"
                },
                {
                    "text": "[I][2021-04-02T19:04:29+0000] Mount: '/lib32' -> '/lib32' flags:MS_RDONLY|MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:true mandatory:false"
                },
                {
                    "text": "[I][2021-04-02T19:04:29+0000] Mount: '/tmp' flags:MS_NOSUID|MS_NODEV|MS_NOEXEC type:'tmpfs' options:'size=1048576,nr_inodes=100' dir:true"
                },
                {
                    "text": "[I][2021-04-02T19:04:29+0000] Mount: '/dev' flags:MS_RDONLY type:'tmpfs' options:'' dir:true"
                },
                {
                    "text": "[I][2021-04-02T19:04:29+0000] Mount: '/dev/null' -> '/dev/null' flags:MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:false"
                },
                {
                    "text": "[I][2021-04-02T19:04:29+0000] Mount: '/dev/zero' -> '/dev/zero' flags:MS_RDONLY|MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:false"
                },
                {
                    "text": "[I][2021-04-02T19:04:29+0000] Mount: '/dev/urandom' -> '/dev/random' flags:MS_RDONLY|MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:false"
                },
                {
                    "text": "[I][2021-04-02T19:04:29+0000] Mount: '/dev/urandom' -> '/dev/urandom' flags:MS_RDONLY|MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:false"
                },
                {
                    "text": "[I][2021-04-02T19:04:29+0000] Mount: '/proc' flags:MS_RDONLY type:'proc' options:'' dir:true"
                },
                {
                    "text": "[I][2021-04-02T19:04:29+0000] Mount: '/opt/compiler-explorer' -> '/opt/compiler-explorer' flags:MS_RDONLY|MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:true"
                },
                {
                    "text": "[I][2021-04-02T19:04:29+0000] Mount: '/opt/intel' -> '/opt/intel' flags:MS_RDONLY|MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:true mandatory:false"
                },
                {
                    "text": "[I][2021-04-02T19:04:29+0000] Mount: '/opt/arm' -> '/opt/arm' flags:MS_RDONLY|MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:true mandatory:false"
                },
                {
                    "text": "[I][2021-04-02T19:04:29+0000] Mount: '/usr/include' -> '/usr/include' flags:MS_RDONLY|MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:true"
                },
                {
                    "text": "[I][2021-04-02T19:04:29+0000] Mount: '/tmp/compiler-explorer-compiler202132-4042-1j96fpe.dxrt' -> '/app' flags:MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:true"
                },
                {
                    "text": "[I][2021-04-02T19:04:29+0000] Uid map: inside_uid:0 outside_uid:116 count:1 newuidmap:false"
                },
                {
                    "text": "[I][2021-04-02T19:04:29+0000] Gid map: inside_gid:0 outside_gid:121 count:1 newgidmap:false"
                },
                {
                    "text": "[I][2021-04-02T19:04:29+0000] Executing '/opt/compiler-explorer/cmake/bin/cmake' for '[STANDALONE MODE]'"
                },
                {
                    "text": "CMake Error: CMake was unable to find a build program corresponding to \"Unix Makefiles\".  CMAKE_MAKE_PROGRAM is not set.  You probably need to select a different build tool."
                },
                {
                    "text": "CMake Error: CMAKE_C_COMPILER not set, after EnableLanguage"
                },
                {
                    "text": "CMake Error: CMAKE_CXX_COMPILER not set, after EnableLanguage"
                },
                {
                    "text": "[I][2021-04-02T19:04:30+0000] pid=5300 ([STANDALONE MODE]) exited with status: 1, (PIDs left: 0)"
                }
            ],
            "execTime": "374",
            "step": "cmake"
        },
        {
            "code": 255,
            "okToCache": true,
            "stdout": [],
            "stderr": [
                {
                    "text": "[I][2021-04-02T19:04:30+0000] Mode: STANDALONE_ONCE"
                },
                {
                    "text": "[I][2021-04-02T19:04:30+0000] Jail parameters: hostname:'ce', chroot:'', process:'make', bind:[::]:0, max_conns_per_ip:0, time_limit:0, personality:0, daemonize:false, clone_newnet:true, clone_newuser:true, clone_newns:true, clone_newpid:true, clone_newipc:true, clone_newuts:true, clone_newcgroup:true, keep_caps:false, disable_no_new_privs:false, max_cpus:1"
                },
                {
                    "text": "[I][2021-04-02T19:04:30+0000] Mount: '/' flags:MS_RDONLY type:'tmpfs' options:'' dir:true"
                },
                {
                    "text": "[I][2021-04-02T19:04:30+0000] Mount: '/lib' -> '/lib' flags:MS_RDONLY|MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:true"
                },
                {
                    "text": "[I][2021-04-02T19:04:30+0000] Mount: '/usr/lib' -> '/usr/lib' flags:MS_RDONLY|MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:true"
                },
                {
                    "text": "[I][2021-04-02T19:04:30+0000] Mount: '/usr/share/zoneinfo' -> '/usr/share/zoneinfo' flags:MS_RDONLY|MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:true"
                },
                {
                    "text": "[I][2021-04-02T19:04:30+0000] Mount: '/etc/localtime' -> '/etc/localtime' flags:MS_RDONLY|MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:false"
                },
                {
                    "text": "[I][2021-04-02T19:04:30+0000] Mount: '/lib64' -> '/lib64' flags:MS_RDONLY|MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:true mandatory:false"
                },
                {
                    "text": "[I][2021-04-02T19:04:30+0000] Mount: '/lib32' -> '/lib32' flags:MS_RDONLY|MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:true mandatory:false"
                },
                {
                    "text": "[I][2021-04-02T19:04:30+0000] Mount: '/tmp' flags:MS_NOSUID|MS_NODEV|MS_NOEXEC type:'tmpfs' options:'size=1048576,nr_inodes=100' dir:true"
                },
                {
                    "text": "[I][2021-04-02T19:04:30+0000] Mount: '/dev' flags:MS_RDONLY type:'tmpfs' options:'' dir:true"
                },
                {
                    "text": "[I][2021-04-02T19:04:30+0000] Mount: '/dev/null' -> '/dev/null' flags:MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:false"
                },
                {
                    "text": "[I][2021-04-02T19:04:30+0000] Mount: '/dev/zero' -> '/dev/zero' flags:MS_RDONLY|MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:false"
                },
                {
                    "text": "[I][2021-04-02T19:04:30+0000] Mount: '/dev/urandom' -> '/dev/random' flags:MS_RDONLY|MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:false"
                },
                {
                    "text": "[I][2021-04-02T19:04:30+0000] Mount: '/dev/urandom' -> '/dev/urandom' flags:MS_RDONLY|MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:false"
                },
                {
                    "text": "[I][2021-04-02T19:04:30+0000] Mount: '/proc' flags:MS_RDONLY type:'proc' options:'' dir:true"
                },
                {
                    "text": "[I][2021-04-02T19:04:30+0000] Mount: '/opt/compiler-explorer' -> '/opt/compiler-explorer' flags:MS_RDONLY|MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:true"
                },
                {
                    "text": "[I][2021-04-02T19:04:30+0000] Mount: '/opt/intel' -> '/opt/intel' flags:MS_RDONLY|MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:true mandatory:false"
                },
                {
                    "text": "[I][2021-04-02T19:04:30+0000] Mount: '/opt/arm' -> '/opt/arm' flags:MS_RDONLY|MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:true mandatory:false"
                },
                {
                    "text": "[I][2021-04-02T19:04:30+0000] Mount: '/usr/include' -> '/usr/include' flags:MS_RDONLY|MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:true"
                },
                {
                    "text": "[I][2021-04-02T19:04:30+0000] Mount: '/tmp/compiler-explorer-compiler202132-4042-1j96fpe.dxrt' -> '/app' flags:MS_BIND|MS_REC|MS_PRIVATE type:'' options:'' dir:true"
                },
                {
                    "text": "[I][2021-04-02T19:04:30+0000] Uid map: inside_uid:0 outside_uid:116 count:1 newuidmap:false"
                },
                {
                    "text": "[I][2021-04-02T19:04:30+0000] Gid map: inside_gid:0 outside_gid:121 count:1 newgidmap:false"
                },
                {
                    "text": "[I][2021-04-02T19:04:30+0000] Executing 'make' for '[STANDALONE MODE]'"
                },
                {
                    "text": "[E][2021-04-02T19:04:30+0000][1] void subproc::subprocNewProc(nsjconf_t*, int, int, int, int, int)():205 execve('make') failed: No such file or directory"
                },
                {
                    "text": "[F][2021-04-02T19:04:30+0000][1] bool subproc::runChild(nsjconf_t*, int, int, int, int)():448 Launching child process failed"
                },
                {
                    "text": "[W][2021-04-02T19:04:30+0000][5309] bool subproc::runChild(nsjconf_t*, int, int, int, int)():478 Received error message from the child process before it has been executed"
                },
                {
                    "text": "[E][2021-04-02T19:04:30+0000][5309] int nsjail::standaloneMode(nsjconf_t*)():256 Couldn't launch the child process"
                }
            ],
            "execTime": "29",
            "step": "make"
        }
    ],
    "result": {
        "okToCache": true,
        "asm": [
            {
                "text": "<No output: objdump returned 1>",
                "source": null
            }
        ],
        "tools": []
    }
}
```


TLDR: `CMake Error: CMake was unable to find a build program corresponding to \"Unix Makefiles\".  CMAKE_MAKE_PROGRAM is not set.  You probably need to select a different build tool."`

So it needs to be able to find `/usr/bin/make`

**Comment by partouf** [comment]
Swift can't link, but unsure about fix (maybe still binutils, but if it's missing it tries this?)
`clang-10: error: invalid linker name in argument '-fuse-ld=gold'`

**Comment by partouf** [comment]
Rust can't link because of unknown error https://godbolt.org/beta/z/3seWWf6c1

**Comment by partouf** [comment]
Python fails because it can't access `/infra/.deploy/etc/scripts`
https://godbolt.org/beta/z/P31KE4rjf

**Comment by partouf** [comment]
Unclear why nim fails, https://godbolt.org/beta/z/x37n5T3Wc
Seems it's needs the system's gcc?

**Comment by partouf** [comment]
Unclear why Haskell fails https://godbolt.org/beta/z/fcoeMq9r9

**Comment by partouf** [comment]
GO fails https://godbolt.org/beta/z/rY9dvK5vz

Maybe because GO requires the environment variables:
```
execOptions.env.GOROOT = goroot;
execOptions.env.GOARCH = goarch;
execOptions.env.GOOS = goos;
```

**Comment by partouf** [comment]
D fails, seems to require system's gcc?
https://godbolt.org/beta/z/qjbGMn7eE

**Comment by mattgodbolt** [comment]
> Python fails because it can't access /infra/.deploy/etc/scripts

Fixed in ceeedef

> https://godbolt.org/beta/z/5bT8Tc1fq -> pascal broken, it's looking for `as`, so system's binutils need to be accessible

Yeah I spotted that too -- looks like a surprising number of things rely on the system binutils et al. I propose we add a PATH to everything to pick up a good, known "stable" binutils/gcc combo.

Thanks for the review comments so far, will pick away at them tomorrow.

File: etc/nsjail/execute.cfg (hunk 2 above)
Comment (mattgodbolt) [reply etc/nsjail/execute.cfg:14]: Yup will bump.

File: etc/nsjail/execute.cfg (hunk 3 above)
Comment (mattgodbolt) [reply etc/nsjail/execute.cfg:7]: Yeah, this will fix the rust linker I think.

File: etc/nsjail/execute.cfg (hunk 5 above)
Comment (mattgodbolt) [reply etc/nsjail/execute.cfg:35]: Oh got it. I'll happily add those in. Just... seems like a good place to start with little.

File: etc/nsjail/execute.cfg (hunk 4 above)
Comment (mattgodbolt) [reply etc/nsjail/execute.cfg:33]: > Upon further review, for this reason, we probably want to remove the max_cpus from our sandbox config too and rely purely on cgroup_cpu_ms_per_sec.

Will do.

File: etc/nsjail/execute.cfg (hunk 5 above)
Comment (mattgodbolt) [reply etc/nsjail/execute.cfg:35]: Reluctant to "just" add `/etc`. Will do something more choice, we were more careful in firejail.

Comment (apmorton) [reply etc/nsjail/execute.cfg:35]: probably should have been clearer on `/etc`; I think we are safe to just give access to `/usr` and `/bin` outright, and we will likely need to find things on a case by case basis in `/etc` that are needed and specifically bind mount (or fabricate with `src_content`).

**Comment by mattgodbolt** [comment]
Will go over this again tomorrow; thanks all!

File: etc/nsjail/execute.cfg (hunk 5 above)
Comment (mattgodbolt) [reply etc/nsjail/execute.cfg:34]: Absolutely. So far I've fabricated /etc/passwd only. `ld.so.conf.d,ld.so.conf` are the other two mentioned in the `firejail` config but I don't know if/when/how those are actually used.

**Comment by mattgodbolt** [comment]
Issues:
* [ ] Pascal: https://godbolt.org/beta/z/5bT8Tc1fq
* [ ] cmake errors
* [ ] Swift link: https://godbolt.org/beta/z/8c741jYYP
* [ ] Rust link https://godbolt.org/beta/z/3seWWf6c1
* [ ] Python https://godbolt.org/beta/z/P31KE4rjf
* [ ] nim https://godbolt.org/beta/z/x37n5T3Wc
* [ ] Haskell https://godbolt.org/beta/z/fcoeMq9r9
* [ ] Go: https://godbolt.org/beta/z/rY9dvK5vz
* [ ] D: https://godbolt.org/beta/z/qjbGMn7eE
* [ ]

**Comment by mattgodbolt** [comment]
The Pascal:
```
[0.184] Searching file /opt/compiler-explorer/fpc-3.2.0.x86_64-linux/bin/x86_64-linux/as... not found
[0.184] Searching file /opt/compiler-explorer/fpc-3.2.0.x86_64-linux/bin/x86_64-linux/AS... not found
[0.188] Searching file /opt/compiler-explorer/fpc-3.2.0.x86_64-linux/lib/fpc/3.2.0/as... not found
[0.191] Searching file /opt/compiler-explorer/fpc-3.2.0.x86_64-linux/lib/fpc/3.2.0/AS... not found
[0.191] Searching file /usr/local/sbin/as... not found
[0.191] Searching file /usr/local/sbin/AS... not found
[0.191] Searching file /usr/local/bin/as... not found
[0.191] Searching file /usr/local/bin/AS... not found
[0.191] Searching file /usr/sbin/as... not found
[0.191] Searching file /usr/sbin/AS... not found
[0.191] Searching file /usr/bin/as... found
[0.191] Using assembler: /usr/bin/as
```
is what it says when it works (on master).

```
[0.152] Assembling output
[0.152] Searching file /opt/compiler-explorer/fpc-3.2.0.x86_64-linux/bin/x86_64-linux/as... not found
[0.152] Searching file /opt/compiler-explorer/fpc-3.2.0.x86_64-linux/bin/x86_64-linux/AS... not found
[0.152] Searching file /opt/compiler-explorer/fpc-3.2.0.x86_64-linux/lib/fpc/3.2.0/as... not found
[0.152] Searching file /opt/compiler-explorer/fpc-3.2.0.x86_64-linux/lib/fpc/3.2.0/AS... not found
[0.152] Searching file as... not found
[0.152] Searching file AS... not found
```

is what it says on the branch. Implies the `PATH` needs to be set up. I suspect that's not making it through to the jail.

**Comment by mattgodbolt** [comment]
When I run just locally: 
```
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

is the path. I wonder why it's not even searching for it in pascal. Perhaps there's something in the specific nsjail cmdline. Will dig, as there's a number of things not finding `as` etc

**Comment by mattgodbolt** [comment]
> When I run just locally:
> 
> ```
> /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
> ```
> 
> is the path. I wonder why it's not even searching for it in pascal. Perhaps there's something in the specific nsjail cmdline. Will dig, as there's a number of things not finding `as` etc

Found it, we don't by default pass any environment. I added a simple `PATH` to find the right things. The firejail has a list of ENVs to remove from the parent env. I wonder if we'll hit other things; we can always pass through but override the banlist.

File: lib/exec.js (hunk 1 above)
Comment (mattgodbolt) [reply lib/exec.js:470]: Yeah I dithered on that. But it sounded dishonest to call firejail from a function called "exeuteNsjail" :grin:

**Comment by partouf** [comment]
Cmake part of cmake api appears to work. Except that it assumes `/usr/bin/g++` is the compiler. It's supposed to get the compiler path from environment variables `CXX`, `CC`, etc (https://github.com/compiler-explorer/compiler-explorer/blob/main/lib/base-compiler.js#L130 and https://github.com/compiler-explorer/compiler-explorer/blob/main/lib/base-compiler.js#L968)

I'm guessing because those are not passed to `.additionalEnv`, but to `.env` of the options https://github.com/compiler-explorer/compiler-explorer/blob/main/lib/exec.js#L180)

Probably the same is happening as with Go (various architectures and WASM don't work because that requires `GOROOT`, `GOOS` and `GOARCH` https://github.com/compiler-explorer/compiler-explorer/blob/main/lib/compilers/golang.js#L224)

**Comment by mattgodbolt** [comment]
Got it; I'll switch to letting all the env vars come through unmolested, execpt those we current blank in the firejail config.

**Comment by mattgodbolt** [comment]
With the exception of the INFO level output, this is ready for merge. Any further thoughts welcomed! And of course, testing!

**Comment by partouf** [comment]
Cmake detects the correct compiler.
Make is broken, but I think that's faulty code, checking...

File: etc/nsjail/execute.cfg [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+mount {
+    src: "/opt/compiler-explorer"
+    dst: "/opt/compiler-explorer"
+    is_bind: true
+}
+
+# Needed for local libc
+mount {
```
Comment (apmorton) [inline etc/nsjail/execute.cfg:140]: I think this is redundant now that /usr is mounted.

**Comment by partouf** [comment]
> Cmake detects the correct compiler.
> Make is broken, but I think that's faulty code, checking...

Fixed

File: lib/exec.js [hunk 7]
```diff
@@ -150,27 +150,28 @@ function executeDirect(command, args, options, filenameTransform) {
     });
 }
 
-function withNsjailTimeout(args, options) {
-    if (options && options.timeoutMs) {
+export function getNsJailOptions(configName, command, args, options) {
+    options = {...(options || {})};
```
Comment (apmorton) [inline lib/exec.js:154]: This can just be `options = {...options}`. Object-spread on undefined is perfectly legal and simply returns an empty object.

File: etc/nsjail/execute.cfg (hunk 6 above)
Comment (mattgodbolt) [reply etc/nsjail/execute.cfg:140]: Good point! fixed.

**Comment by mattgodbolt** [comment]
Thanks both. I'm going to merge now, but leave it beta-only, with a view to updating it tomorrow sometime (leaving beta running for folks to take potshots)

File: lib/exec.js [hunk 8]
```diff
@@ -179,29 +180,43 @@ function sandboxNsjail(command, args, options) {
...
+        options,
+        filenameTransform,
+    };
+}
 
+function sandboxNsjail(command, args, options) {
+    logger.info('Sandbox execution via nsjail', {command, args});
+    const execPath = path.dirname(command);
+    const execName = path.basename(command);
+    options = options || {};
```
Comment (apmorton) [inline lib/exec.js:196]: this should be `options = {...options, customCwd: execPath}` to avoid modifying the options object being passed into `sandboxNsjail`

File: lib/exec.js (hunk 7 above)
Comment (mattgodbolt) [reply lib/exec.js:154]: oh cool! thanks!

File: lib/exec.js (hunk 8 above)
Comment (mattgodbolt) [reply lib/exec.js:196]: Done, thanks.

**Comment by mattgodbolt** [comment]
Needs more testing post #2571 review and merge.

And also need to test with tools, PVS Studio etc

* [ ] PVS Studio
* [ ] clang-tidy
* [ ] llvm-mca
* [ ] OSACA
* [ ] pahole
* [ ] clang-format
* [ ] clang-query
* [ ] readelf
* [ ] x86-to-6502
* [ ] ldd
* [ ] include-what-you-use
* [ ] strings

For completeness I guess also:
* [ ] GCC tree view
* [ ] Graph output
* [ ] AST output
* [ ] Opt output
* [ ] IR output

**Comment by partouf** [comment]
CMake still works.
Disappointment from GO https://godbolt.org/beta/z/Tq4Tj1GfY

**Comment by mattgodbolt** [comment]
Oh blast re: `go`, it seems to be looking at the wrong $HOME dir. will fix. Didn't get the time I hoped today so far.

**Comment by mattgodbolt** [comment]
I'm wondering if I forgot to push from me desktop computer. argh

**Comment by mattgodbolt** [comment]
Oh no, I just forgot I hadn't finished all the changes...

**Comment by mattgodbolt** [comment]
Go now works (the link you provided fails for lack of a `main` function but seems to be consistent with what the main site does. Will check the other things. I also made this INFO level, so I think this is in principal mergable as-is.

**Comment by mattgodbolt** [comment]
https://godbolt.org/beta/z/hzh43eK9M - all the trimmings check out. I think this is good to go. @apmorton / @partouf would appreciate a final go-over and then I"ll merge. NB we still haven't enabled this for `amazon` (just `beta`) so we can deploy this then have a separate toggle-it-on change (makes it easier to revert just the one)
